python/ray/train/callbacks/callback.py

```diff
@@ -5,16 +5,8 @@
 class TrainingCallback(abc.ABC):
     """Abstract Train callback class."""
 
-    def handle_result(self, results: List[Dict], **info):
-        """Called every time train.report() is called.
-
-        Args:
-            results (List[Dict]): List of results from the training
-                function. Each value in the list corresponds to the output of
-                the training function from each worker.
-            **info: kwargs dict for forward compatibility.
-        """
-        pass
+    def __init__(self):
+        self._results_preprocessors = []
```

Comment:
We'll have to make sure that all callbacks call `super()`/initialize that variable itself. That could lead to unexpected issues, especially with user-defined callbacks. Would be great if we could avoid that. Maybe we can make it a property somehow?

Reply:
Oh yeah I was considering using `getattr` for this, but was not sure if there was a better way to do it.

Reply:
Maybe one of those (depending on if you prefer `getattr` or exceptions)
```python
    @property
    def results_preprocessors(self):
        self._results_preprocessors = getattr(self, "_results_preprocessors", [])
        return self._results_preprocessors

    @property
    def results_preprocessors(self):
        try:
            _results_preprocessors = self._results_preprocessors
        except AttributeError:
            self._results_preprocessors = []
            _results_preprocessors = self._results_preprocessors
        return _results_preprocessors
```

Reply:
Alternatively, consider using a Mixin for callbacks that use preprocessors.

Reply:
Another option would be to add `_results_preprocessors` as a class variable and override with an instance variable if needed by a callback. I think this should be a clean solution, avoiding potentially confusing mixins and super() calls